src/Pipeline.h

```diff
@@ -437,29 +438,86 @@ struct make_argument_list<void> {
 
 
 template <typename... Args>
-void init_arg_types(std::vector<ScalarOrBufferT> &arg_types) {
+std::vector<ScalarOrBufferT> init_arg_types() {
+    std::vector<ScalarOrBufferT> arg_types;
     make_argument_list<Args..., void>::add_args(arg_types);
+    return arg_types;
 }
 
-}
+}  // namespace
 
-struct JITExtern {
-    // assert pipeline.defined() == (c_function == nullptr) -- strictly one or the other
-    // which should be enforced by the constructors.
-    Pipeline pipeline;
+struct ExternSignature {
+private:
+    Type ret_type_;       // Only meaningful if is_void_return is false; must be default value otherwise
+    bool is_void_return_{false};
+    std::vector<ScalarOrBufferT> arg_types_;
+
+public:
+    ExternSignature() = default;
+
+    ExternSignature(const Type &ret_type, bool is_void_return, const std::vector<ScalarOrBufferT> &arg_types)
+        : ret_type_(ret_type)
+        , is_void_return_(is_void_return)
+        , arg_types_(arg_types) {
+        internal_assert(!(is_void_return && ret_type != Type())); 
+    }
 
-    void *c_function;
-    ExternSignature signature;
+    template <typename RT, typename... Args>
+    ExternSignature(RT (*f)(Args... args)) 
+        : ret_type_(type_of<RT>())
+        , is_void_return_(std::is_void<RT>::value)
```

Comment:
I think other code that breaks up initializers per-line puts the commas at the end of each line rather than the start.